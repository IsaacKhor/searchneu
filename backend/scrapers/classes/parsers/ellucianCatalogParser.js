/*
 * Copyright (c) 2017 Ryan Hughes
 *
 * This file is part of CoursePro.
 *
 * CoursePro is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License
 * version 3 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>. 
 */

import URI from 'urijs';
import domutils from 'domutils';
import he from 'he';
import _ from 'lodash';
import cheerio from 'cheerio';


import cache from '../../cache';
import macros from '../../../macros';
import Request from '../../request';

var EllucianBaseParser = require('./ellucianBaseParser').EllucianBaseParser;
var ellucianClassParser = require('./ellucianClassParser');
var ellucianRequisitesParser = require('./ellucianRequisitesParser');
var ellucianRequisitesParser2 = require('./ellucianRequisitesParser2');


const request = new Request('EllucianCatalogParser');

function EllucianCatalogParser() {
	EllucianBaseParser.prototype.constructor.apply(this, arguments);
	this.name = "EllucianCatalogParser"
}


//prototype constructor
EllucianCatalogParser.prototype = Object.create(EllucianBaseParser.prototype);
EllucianCatalogParser.prototype.constructor = EllucianCatalogParser;



EllucianCatalogParser.prototype.supportsPage = function (url) {
	return url.indexOf('bwckctlg.p_disp_course_detail') > -1;
}


// 
EllucianCatalogParser.prototype.main = async function(url) {

  // Possibly load from DEV
  if (macros.DEV && require.main !== module) {
    const devData = await cache.get('dev_data', this.constructor.name, url);
    if (devData) {
      return devData;
    }
  }

  let resp = await request.get(url);

  let retVal = await this.parse(resp.body, url)

 // Possibly save to dev
  if (macros.DEV && require.main !== module) {
    await cache.set('dev_data', this.constructor.name, url, retVal);

    // Don't log anything because there would just be too much logging. 
  }

  return retVal

};


EllucianCatalogParser.prototype.parseClass = function (element, url) {


	let {termId, classId, subject} = this.parseCatalogUrl(url);


	var depData = {
		desc: '',
		classId: classId
	};



	depData.prettyUrl = url

	//get the class name
	var value = domutils.getText(element);

	var match = value.match(/.+?\s-\s*(.+)/i);
	if (!match || match.length < 2 || match[1].length < 2) {
		macros.error('could not find title!', match, value, url);
		return;
	}
	depData.name = this.standardizeClassName(match[1]);


	//find the box below this row
	var descTR = element.parent.next
	while (descTR.type != 'tag') {
		descTR = descTR.next;
	}
	var rows = domutils.getElementsByTagName('td', descTR)
	if (rows.length != 1) {
		macros.error('td rows !=1??', depData.classId, url);
		return;
	};

	element = rows[0]


	//get credits from element.parent.text here

	//grab credits
	var text = domutils.getText(element.parent)
	var creditsParsed = this.parseCredits(text);

	if (creditsParsed) {
		depData.maxCredits = creditsParsed.maxCredits;
		depData.minCredits = creditsParsed.minCredits;
	}
	else {
		macros.log('warning, nothing matchied credits', url, text);
	}


	//desc
	//list all texts between this and next element, not including <br> or <i>
	//usally stop at <span> or <p>
	for (var i = 0; i < element.children.length; i++) {
		if (element.children[i].type == 'tag' && !['i', 'br', 'a'].includes(element.children[i].name)) {
			break;
		}
		depData.desc += '  ' + domutils.getText(element.children[i]).trim();
	}

	depData.desc = depData.desc.replace(/\n|\r/gi, ' ').trim()
		//remove credit hours
		// 0.000 TO 1.000 Credit hours
	depData.desc = depData.desc.replace(/(\d+(\.\d+)?\s+TO\s+)?\d+(.\d+)?\s+credit hours/gi, '').trim();

	depData.desc = depData.desc.replace(/\s+/gi, ' ').trim()

	var invalidDescriptions = ['xml extract', 'new search'];

	if (invalidDescriptions.includes(depData.desc.trim().toLowerCase())) {
		return;
	}



	//url
	depData.url = this.createClassURL(url, termId, subject, classId);
	if (!depData.url) {
		macros.log('error could not create class url', depData);
		return;
	}

	let fakePageData = {
		dbData: {
			url: url,
			termId: termId
		}
	}

	//find co and pre reqs and restrictions
	var prereqs = ellucianRequisitesParser.parseRequirementSection(fakePageData, element.children, 'prerequisites');
	if (prereqs) {
		depData.prereqs = prereqs;
	}

	var coreqs = ellucianRequisitesParser.parseRequirementSection(fakePageData, element.children, 'corequisites');
	if (coreqs) {
		depData.coreqs = coreqs;
	}
	
	//find co and pre reqs and restrictions
	var prereqs2 = ellucianRequisitesParser2.parseRequirementSection(fakePageData, element.children, 'prerequisites');
	if (!_.isEqual(prereqs, prereqs2)) {
		macros.log("WARNING: prereqs parsed by the new parser are not equal", JSON.stringify(prereqs, null, 4), JSON.stringify(prereqs2, null, 4))
	}

	var coreqs2 = ellucianRequisitesParser2.parseRequirementSection(fakePageData, element.children, 'corequisites');
	if (!_.isEqual(coreqs, coreqs2)) {
		macros.log("WARNING: coreqs parsed by the new parser are not equal", JSON.stringify(coreqs, null, 4), JSON.stringify(coreqs2, null, 4))
	}

	console.log(depData)

	// var dep = pageData.addDep(depData);
	// dep.setParser(ellucianClassParser)
};


EllucianCatalogParser.prototype.parse = function (body, url) {

	// Parse the dom
	const $ = cheerio.load(body);

	let elements = $('td.nttitle')

	let matchingElement;

	for (var i = 0; i < elements.length; i++) {
		let currElement = elements[i];

		if (matchingElement) {
			macros.error("Already have a matching element", elements, elements.length)
		}

		if (currElement.parent.parent.attribs.summary.includes('term')) {
			matchingElement = currElement;
		}
	}


	this.parseClass(matchingElement, url);
};







EllucianCatalogParser.prototype.EllucianCatalogParser = EllucianCatalogParser;
module.exports = new EllucianCatalogParser();


function testFunc() {
	module.exports.main('https://wl11gp.neu.edu/udcprod8/bwckctlg.p_disp_course_detail?cat_term_in=201810&subj_code_in=FINA&crse_numb_in=6283')
}


if (require.main === module) {
	testFunc()
	// module.exports.tests();
}
