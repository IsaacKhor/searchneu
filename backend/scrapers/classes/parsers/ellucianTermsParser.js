/*
 * Copyright (c) 2017 Ryan Hughes
 *
 * This file is part of CoursePro.
 *
 * CoursePro is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License
 * version 3 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>. 
 */

import moment from 'moment';
import cheerio from 'cheerio';

import cache from '../../cache';
import macros from '../../../macros';
import Request from '../../request';
var collegeNamesParser = require('./collegeNamesParser');

var EllucianBaseParser = require('./ellucianBaseParser').EllucianBaseParser;
var ellucianSubjectParser = require('./ellucianSubjectParser');


const request = new Request('EllucianTermsParser');


function EllucianTermsParser() {
  EllucianBaseParser.prototype.constructor.apply(this, arguments);
  this.name = "EllucianTermsParser";
}



//prototype constructor
EllucianTermsParser.prototype = Object.create(EllucianBaseParser.prototype);
EllucianTermsParser.prototype.constructor = EllucianTermsParser;



EllucianTermsParser.prototype.getDataType = function (pageData) {
  return 'terms';
};


EllucianTermsParser.prototype.supportsPage = function (url) {
  return url.indexOf('bwckschd.p_disp_dyn_sched') > -1;
};


// 
EllucianTermsParser.prototype.main = async function(url) {
  
  // Possibly load from DEV
  if (macros.DEV && require.main !== module) {
    const devData = await cache.get('dev_data', this.constructor.name, url);
    if (devData) {
      return devData;
    }
  }

  let resp = await request.get(url);

  let retVal = this.parse(resp.body, url)

 // Possibly save to dev
  if (macros.DEV && require.main !== module) {
    await cache.set('dev_data', this.constructor.name, url, retVal);

    // Don't log anything because there would just be too much logging. 
  }

  return retVal

};



EllucianTermsParser.prototype.minYear = function () {
  return moment().subtract(4, 'months').year()
};

EllucianTermsParser.prototype.isValidTerm = function (termId, text) {

  var year = text.match(/\d{4}/);
  var minYear = this.minYear();

  if (!year) {
    macros.log('warning: could not find year for ', text);

    //if the termId starts with the >= current year, then go
    var idYear = parseInt(termId.slice(0, 4), 10)

    //if first 4 numbers of id are within 3 years of the year that it was 4 months ago
    if (idYear + 3 > minYear && idYear - 3 < minYear) {
      return true;
    }
    else {
      return false;
    }
  }

  //skip past years
  if (parseInt(year, 10) < minYear) {
    return false;
  }
  return true;

};



EllucianTermsParser.prototype.parse = async function (body, url) {
  var formData = this.parseTermsPage(body, url);
  var terms = [];

  formData.requestsData.forEach(function (singleRequestPayload) {

    //record all the terms and their id's
    singleRequestPayload.forEach(function (payloadVar) {
      if (this.shouldParseEntry(payloadVar)) {
        terms.push({
          id: payloadVar.value,
          text: payloadVar.text
        });
      }
    }.bind(this));
  }.bind(this));

  if (terms.length === 0) {
    macros.log('ERROR, found 0 terms??', url);
  };

  var host = macros.getBaseHost(url);

  terms.forEach(function (term) {

    //calculate host for each entry
    var host = macros.getBaseHost(url);

    // If this is a term that matches a term in staticHosts
    // Remove 
    let possibleCustomHostAndText = collegeNamesParser.getHostForTermTitle(host, term.text);

    if (possibleCustomHostAndText) {
      term.text = possibleCustomHostAndText.text
      term.host = possibleCustomHostAndText.host
    }

    //add the shorter version of the term string
    term.shortText = term.text.replace(/Quarter|Semester/gi, '').trim()

  }.bind(this))



  let duplicateTexts = {};



  //keep track of texts, and if they are all different with some words removed
  //keep the words out
  terms.forEach(function (term) {

    if (!duplicateTexts[term.host]) {
      duplicateTexts[term.host] = {
        values: [],
        areAllDifferent: true
      }
    }
    if (duplicateTexts[term.host].values.includes(term.shortText)) {
      duplicateTexts[term.host].areAllDifferent = false;
      return;
    }
    duplicateTexts[term.host].values.push(term.shortText)

  }.bind(this))


  //for each host, change the values if they are different
  terms.forEach(function (term) {
    if (duplicateTexts[term.host].areAllDifferent) {
      term.text = term.shortText
    }
  }.bind(this))




  let outputTerms = []


  //the given page data is the controller
  //give the first term to it,
  //and pass the others in as deps + noupdate


  let subjectPromises = {}


  terms.forEach(function (term) {

    macros.log("Parsing term: ", JSON.stringify(term));

    // Parse the subjects. Keep track of all the promises in a map. 
    subjectPromises[term.id] = ellucianSubjectParser.main(formData.postURL, term.id)

  }.bind(this))


  // Wait for all the subjects to be parsed.
  await Promise.all(Object.values(subjectPromises));

  for (const term of terms) {
    outputTerms.push({
      termId: term.id,
      text: term.text,
      host: term.host,
      subjects: await subjectPromises[term.id]
    })
  }

  return outputTerms
};


EllucianTermsParser.prototype.shouldParseEntry = function(entry) {
  if (entry.name == 'p_term'){
    return true;
  }
  else {
    return false;
  }
};



//step 1, select the terms
//starting url is the terms page
EllucianTermsParser.prototype.parseTermsPage = function (body, url) {

  // Parse the dom
  const $ = cheerio.load(body);

  var parsedForm = this.parseForm(url, $('body')[0]);

  if (!parsedForm) {
    macros.error('default form data failed');
    return;
  }

  var defaultFormData = parsedForm.payloads;


  //find the term entry and all the other entries
  var termEntry;
  var otherEntries = [];
  defaultFormData.forEach(function (entry) {
    if (this.shouldParseEntry(entry)) {
      if (termEntry) {
        macros.error("Already and entry???", termEntry)
      }
      termEntry = entry;
    }
    else {
      otherEntries.push(entry);
    }
  }.bind(this));

  if (!termEntry) {
    macros.error('Could not find an entry!', url, JSON.stringify(parsedForm));
    return;
  }

  var requestsData = []; 

  //setup an indidual request for each valid entry on the form - includes the term entry and all other other entries
  termEntry.alts.forEach(function (entry) {
    if (!this.shouldParseEntry(entry)) {
      macros.log('ERROR: entry was alt of term entry but not same name?', entry);
      return;
    }
    entry.text = entry.text.trim()

    if (entry.text.toLowerCase() === 'none') {
      return;
    }
    entry.text = entry.text.replace(/\(view only\)/gi, '').trim();

    entry.text = entry.text.replace(/summer i$/gi, 'Summer 1').replace(/summer ii$/gi, 'Summer 2')

    //dont process this element on error
    if (entry.text.length < 2) {
      macros.log('warning: empty entry.text on form?', entry, url);
      return;
    }

    if (!this.isValidTerm(entry.value, entry.text)) {
      return;
    }


    var fullRequestData = otherEntries.slice(0);

    fullRequestData.push({
      name: entry.name,
      value: entry.value,
      text: entry.text
    });

    requestsData.push(fullRequestData);

  }.bind(this));

  return {
    postURL: parsedForm.postURL,
    requestsData: requestsData
  };
};




EllucianTermsParser.prototype.EllucianTermsParser = EllucianTermsParser;
module.exports = new EllucianTermsParser();


async function testFunc() {
  let r = await module.exports.main('https://wl11gp.neu.edu/udcprod8/bwckschd.p_disp_dyn_sched')


  console.log(r)
}


if (require.main === module) {
  testFunc();
}
