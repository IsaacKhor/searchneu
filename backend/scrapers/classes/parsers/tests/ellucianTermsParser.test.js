/*
 * Copyright (c) 2017 Ryan Hughes
 *
 * This file is part of CoursePro.
 *
 * CoursePro is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License
 * version 3 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>. 
 */

import path from 'path';
import MockDate from 'mockdate';
import fs from 'fs-promise';
import URI from 'urijs';
import cheerio from 'cheerio';

import ellucianCatalogParser from '../ellucianCatalogParser';
import ellucianTermsParser from '../ellucianTermsParser';


beforeAll(function () {
  MockDate.set('10/10/2016');
});

afterAll(function() {
  MockDate.reset();
});



it('has a name', function() {
  
  //make sure a name is defined
  expect(ellucianTermsParser.name);
});

it('isValidTerm should work', function () {

  expect(ellucianTermsParser.isValidTerm('201630', 'blah blah 2016')).toBe(true)
  expect(ellucianTermsParser.isValidTerm('201630', 'blah blah 2017')).toBe(true)
  expect(ellucianTermsParser.isValidTerm('201630', 'blah blah')).toBe(true)
  expect(ellucianTermsParser.isValidTerm('2016', 'blah blah')).toBe(true)
  expect(ellucianTermsParser.isValidTerm('201', 'blah blah')).toBe(false)
});
  
it('should behave...', async function (done) {

  const body = await fs.readFile(path.join(__dirname, 'data', 'ellucianTermsParser', '1.html'), 'utf8');

  var url = 'https://bannerweb.upstate.edu/isis/bwckschd.p_disp_dyn_sched';
  expect(true).toBe(ellucianTermsParser.supportsPage(url));

  expect(ellucianTermsParser.parse(body, url)).toMatchSnapshot();

  done()
});



it('hi there 2', async function (done) {

  const body = await fs.readFile(path.join(__dirname, 'data', 'ellucianTermsParser', '2.html'), 'utf8');

  var url = 'https://wl11gp.neu.edu/udcprod8/bwckschd.p_disp_dyn_sched';

  let output = ellucianTermsParser.parse(body, url);

  expect(true).toBe(ellucianTermsParser.supportsPage(url));

  expect(output.terms.length).toBe(17);

  done()
});

