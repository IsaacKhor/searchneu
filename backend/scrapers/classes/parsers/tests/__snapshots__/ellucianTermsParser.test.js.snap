// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should behave... 1`] = `
Object {
  "postUrl": "https://bannerweb.upstate.edu/isis/bwckgens.p_proc_term_date",
  "terms": Array [
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "upstate.edu",
        "termId": "201611",
        "text": "Spring 2017 Summer 2",
      },
    },
    Object {
      "deps": null,
      "type": "terms",
      "value": Object {
        "host": "upstate.edu",
        "termId": "201610",
        "text": "Spring 2016",
      },
    },
  ],
}
`;
